# Indian E-Commerce Sales Analysis
## Notebook 01: Data Understanding

**Goal:** Load the three tables (sales, customers, products) and understand their size, structure, and data quality before cleaning.

In [81]:
import pandas as pd
import numpy as np

In [82]:
sales = pd.read_csv("../data/raw/sales.csv")
customers = pd.read_csv("../data/raw/customers.csv")
products = pd.read_csv("../data/raw/products.csv")

In [83]:
print("Sales:", sales.shape)
print("Customers:", customers.shape)
print("Products:", products.shape)

Sales: (250000, 21)
Customers: (40000, 15)
Products: (2000, 12)


**Observation:** All three files loaded successfully. Sales has 250,000 rows, customers 40,000, and products 2,000.

## 1. First Look at the Data

In [84]:
pd.set_option("display.max_columns", None)

In [85]:
sales.head()

,Order_ID,Customer_ID,Product_ID,Order_Date,Order_Time,Delivery_Date,Quantity,Unit_Price,Order_Value,Shipping_Cost,Coupon_Code,Coupon_Discount,Total_Amount,Payment_Mode,Order_Status,Rating,Review_Text,City,State,Customer_Age,Customer_Age_Group
0,ORD0000000001,CUST00014303,PROD001017,2026-06-07,08:20:00,2026-06-10,1,2622.46,2622.46,0.0,NaN,0.0,2622.46,COD,Delivered,NaN,NaN,Chennai,Tamil Nadu,43,36-45
1,ORD0000000002,CUST00034253,PROD000703,2025-02-04,20:05:00,2025-02-07,1,90258.00,90258.00,0.0,NaN,0.0,90258.00,COD,Delivered,4.0,Nice quality,Durgapur,West Bengal,22,18-25
2,ORD0000000003,CUST00000421,PROD000318,2026-03-12,14:59:00,2026-03-19,1,3536.50,3536.50,0.0,NaN,0.0,3536.50,UPI,Processing,NaN,NaN,Patiala,Punjab,41,36-45
3,ORD0000000004,CUST00026507,PROD000213,2025-01-27,14:33:00,2025-02-02,1,34081.56,34081.56,0.0,FLAT50,50.0,34031.56,Credit Card,Delivered,NaN,NaN,Rajkot,Gujarat,29,26-35
4,ORD0000000005,CUST00009418,PROD001437,2024-06-30,07:19:00,2024-07-02,2,69021.17,138042.34,0.0,NaN,0.0,138042.34,UPI,Delivered,NaN,NaN,Surat,Gujarat,39,36-45


In [86]:
display(customers.head())
display(products.head())

,Customer_ID,Customer_Name,Gender,Age,Age_Group,Date_of_Birth,Email,Phone,City,State,Pincode,Registration_Date,Customer_Tier,Total_Orders,Total_Spent
0,CUST00000001,Akash Agarwal,male,48,46-55,1977-01-13,akashagarwal@gmail.com,+91-9625152011,Patiala,Punjab,122521,2024-01-28,Gold,5,40098.48
1,CUST00000002,Farhan Rahman,male,73,65+,1952-01-20,farhanrahman@gmail.com,+91-8593009465,Durgapur,West Bengal,226518,2023-08-25,Platinum,8,63676.24
2,CUST00000003,Suresh Kumar,male,31,26-35,1994-01-09,sureshkumar755@gmail.com,+91-7935884774,Gurugram,Haryana,520400,2023-08-30,Silver,0,0.00
3,CUST00000004,Aditya Pandey,male,41,36-45,1984-01-12,adityapandey@outlook.com,+91-9015046072,Kolkata,West Bengal,735356,2024-03-11,Platinum,5,288324.84
4,CUST00000005,Hina Khan,female,53,46-55,1972-01-15,hinakhan433@gmail.com,+91-8813094271,Noida,UP,985141,2024-04-19,Platinum,4,69265.75


,Product_ID,Product_Name,Category,Brand,Original_Price,Discount_Percent,Discount_Amount,Selling_Price,Stock_Quantity,Weight_kg,Avg_Rating,Total_Reviews
0,PROD000001,Samsung Mobile V5,Electronics,Samsung,130561.34,23,30029.11,100532.23,159,2.63,4.49,102
1,PROD000002,HP Laptop V5,Electronics,HP,36494.92,34,12408.27,24086.65,54,4.19,4.32,107
2,PROD000003,boAt Headphones V2,Electronics,boAt,106587.94,42,44766.93,61821.01,638,1.48,4.31,99
3,PROD000004,Noise Watch V7,Electronics,Noise,81275.27,47,38199.38,43075.89,572,3.30,4.42,127
4,PROD000005,Apple iPhone V4,Electronics,Apple,13099.36,43,5632.72,7466.64,865,3.71,4.37,98


**Observation:** Each row in sales appears to be one order containing one product. Dates are in ISO format. Customer_ID and Product_ID link sales to the other two tables.

## 2. Data Types

In [87]:
sales.info()

<class 'pandas.DataFrame'>
RangeIndex: 250000 entries, 0 to 249999
Data columns (total 21 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Order_ID            250000 non-null  str    
 1   Customer_ID         250000 non-null  str    
 2   Product_ID          250000 non-null  str    
 3   Order_Date          250000 non-null  str    
 4   Order_Time          250000 non-null  str    
 5   Delivery_Date       250000 non-null  str    
 6   Quantity            250000 non-null  int64  
 7   Unit_Price          250000 non-null  float64
 8   Order_Value         250000 non-null  float64
 9   Shipping_Cost       250000 non-null  float64
 10  Coupon_Code         50185 non-null   str    
 11  Coupon_Discount     250000 non-null  float64
 12  Total_Amount        250000 non-null  float64
 13  Payment_Mode        250000 non-null  str    
 14  Order_Status        250000 non-null  str    
 15  Rating              120030 non-null  float64


**Observation (sales.info):**
- 250,000 rows and 21 columns, matching the dataset description.
- Money columns are already numeric, so no currency symbols need removing.
- Order_Date, Order_Time and Delivery_Date are stored as text and must be converted to dates during cleaning.
- Coupon_Code is ~80% empty, most likely because most orders did not use a coupon (to verify: Coupon_Discount should be 0 for these rows).
- Rating and Review_Text are both missing in exactly 129,970 rows (~52%), suggesting a rating and review are always given together. Possible reason: only delivered orders are reviewed (to verify later).

In [88]:
customers.info()
print("-" * 50)
products.info()

<class 'pandas.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Customer_ID        40000 non-null  str    
 1   Customer_Name      40000 non-null  str    
 2   Gender             40000 non-null  str    
 3   Age                40000 non-null  int64  
 4   Age_Group          40000 non-null  str    
 5   Date_of_Birth      40000 non-null  str    
 6   Email              40000 non-null  str    
 7   Phone              40000 non-null  str    
 8   City               40000 non-null  str    
 9   State              40000 non-null  str    
 10  Pincode            40000 non-null  int64  
 11  Registration_Date  40000 non-null  str    
 12  Customer_Tier      40000 non-null  str    
 13  Total_Orders       40000 non-null  int64  
 14  Total_Spent        40000 non-null  float64
dtypes: float64(1), int64(3), str(11)
memory usage: 8.8 MB
---------------------------

## 3. Missing Values

In [89]:
missing = pd.DataFrame({
    "Missing Count": sales.isnull().sum(),
    "Missing %": (sales.isnull().sum() / len(sales) * 100).round(2)
})
missing[missing["Missing Count"] > 0]

,Missing Count,Missing %
Coupon_Code,199815,79.93
Rating,129970,51.99
Review_Text,129970,51.99


In [90]:
print("Customers:\n", customers.isnull().sum()[customers.isnull().sum() > 0])
print("\nProducts:\n", products.isnull().sum()[products.isnull().sum() > 0])

Customers:
 Series([], dtype: int64)

Products:
 Series([], dtype: int64)


## 4. Duplicates

In [91]:
print("Duplicate rows in sales:", sales.duplicated().sum())
print("Duplicate rows in customers:", customers.duplicated().sum())
print("Duplicate rows in products:", products.duplicated().sum())
print("Repeated Order_IDs in sales:", sales["Order_ID"].duplicated().sum())

Duplicate rows in sales: 0
Duplicate rows in customers: 0
Duplicate rows in products: 0
Repeated Order_IDs in sales: 0


**Observation:** Order_Date and Delivery_Date are stored as text and need converting to dates. Coupon_Code has __% missing, which likely means no coupon was used, not an error. No duplicate rows were found.

## 5. Summary Statistics

In [92]:
sales.describe().round(2)

,Quantity,Unit_Price,Order_Value,Shipping_Cost,Coupon_Discount,Total_Amount,Rating,Customer_Age
count,250000.00,250000.00,250000.00,250000.00,250000.00,250000.00,120030.00,250000.00
mean,1.25,19202.48,23967.86,4.91,250.00,23722.77,4.40,35.31
std,0.54,26720.50,37629.32,18.14,1399.93,37267.64,0.66,12.57
min,1.00,26.13,26.13,0.00,0.00,-21.56,3.00,18.00
25%,1.00,1551.55,1825.89,0.00,0.00,1803.90,4.00,26.00
50%,1.00,7208.93,8411.16,0.00,0.00,8339.49,4.00,33.00
75%,1.00,24433.27,28612.81,0.00,0.00,28344.92,5.00,43.00
max,3.00,136973.33,410919.99,90.00,41092.00,410919.99,5.00,75.00


In [93]:
negative = sales[sales["Total_Amount"] < 0]
print("Orders with negative Total_Amount:", len(negative))
print("Orders where coupon > order value:", (sales["Coupon_Discount"] > sales["Order_Value"]).sum())

negative[["Order_ID", "Order_Value", "Shipping_Cost", "Coupon_Code", "Coupon_Discount", "Total_Amount"]].head(10)

Orders with negative Total_Amount: 5
Orders where coupon > order value: 107


,Order_ID,Order_Value,Shipping_Cost,Coupon_Code,Coupon_Discount,Total_Amount
84204,ORD0000084205,35.57,62.36,DIWALI100,100.0,-2.07
182883,ORD0000182884,26.13,52.31,DIWALI100,100.0,-21.56
207695,ORD0000207696,26.13,62.56,DIWALI100,100.0,-11.31
216234,ORD0000216235,26.13,57.40,DIWALI100,100.0,-16.47
249820,ORD0000249821,35.57,52.45,DIWALI100,100.0,-11.98


In [94]:
over_discount = sales["Coupon_Discount"] > sales["Order_Value"]
sales.loc[over_discount, "Coupon_Code"].value_counts()

Coupon_Code
DIWALI100    81
FLAT50       26
Name: count, dtype: int64

In [95]:
sales["Shipping_Cost"].value_counts().head(10)

Shipping_Cost
0.00     232459
78.81        14
70.35        14
57.23        13
77.46        13
53.54        12
55.18        12
55.57        11
57.59        11
69.84        11
Name: count, dtype: int64

**Observation (negative totals):**
- 5 orders (0.002%) have a negative Total_Amount. All used DIWALI100, a flat ₹100 coupon, on items costing ₹26–36.
- 107 orders have a coupon discount larger than the order value; in most, shipping charges kept the total above zero.
- Manual check confirms: Total_Amount = Order_Value + Shipping_Cost − Coupon_Discount.
- Business insight: flat coupons appear to have no minimum order value, creating loss-making orders.
- Cleaning decision: cap Total_Amount at 0 and add a flag column (a checkout never charges below ₹0; revenue impact ≈ ₹63).

In [96]:
paid_shipping = sales["Shipping_Cost"] > 0

print("Orders that paid shipping:", paid_shipping.sum())
print("Highest order value that PAID shipping:", sales.loc[paid_shipping, "Order_Value"].max())
print("Lowest order value with FREE shipping:", sales.loc[~paid_shipping, "Order_Value"].min())

Orders that paid shipping: 17541
Highest order value that PAID shipping: 499.34
Lowest order value with FREE shipping: 500.22


- Free-shipping threshold test: every order below ₹499.34 paid shipping, and every order from ₹500.22 upward shipped free. This shows a clear business rule: free shipping on orders above ~₹500. About 7% of orders (17,541) fall below it.

## 6. Categorical Columns

In [97]:
for col in ["Payment_Mode", "Order_Status", "Customer_Age_Group"]:
    print(sales[col].value_counts(), "\n")

Payment_Mode
UPI            128474
COD             82093
Debit Card      28856
Credit Card     10577
Name: count, dtype: int64 

Order_Status
Delivered     200139
Cancelled      12507
Returned       12493
Shipped        12459
Processing     12402
Name: count, dtype: int64 

Customer_Age_Group
26-35    87847
18-25    61594
36-45    50169
46-55    30359
56-65    15176
65+       4855
Name: count, dtype: int64 



In [98]:
print("Unique states:", sales["State"].nunique())
print("Unique cities:", sales["City"].nunique())
print(sales["State"].value_counts().head(10))

Unique states: 10
Unique cities: 39
State
UP             32631
Haryana        32064
Rajasthan      31666
Maharashtra    25478
Punjab         25235
Gujarat        25171
Tamil Nadu     19792
Delhi          19524
West Bengal    19450
Karnataka      18989
Name: count, dtype: int64


- Payment modes: UPI ~51%, COD ~33%, Debit Card ~12%, Credit Card ~4%. One-third of orders are COD, which is worth testing against cancellations and returns.
- Order statuses: Delivered ~80%; Cancelled, Returned, Shipped and Processing ~5% each. The unusually even split suggests simulated data.
- Revenue definition: GMV = all orders; Net Revenue = Delivered orders only.
- ~60% of orders come from customers aged 18–35.

## 7. Date Range

In [99]:
order_dates = pd.to_datetime(sales["Order_Date"], errors="coerce")
print("First order:", order_dates.min())
print("Last order:", order_dates.max())
print("Dates that could not be read:", order_dates.isnull().sum())

First order: 2024-06-01 00:00:00
Last order: 2026-06-30 00:00:00
Dates that could not be read: 0


**Observation (date range):**
- Orders run from 1 June 2024 to 30 June 2026 (25 months); all dates were read successfully.
- The period includes two Diwali seasons, allowing festive patterns to be checked across both years.
- June appears in three years while other months appear in two, so monthly analysis must group by year-month (or use averages), not by month name.

In [100]:
status_dates = pd.DataFrame({"Status": sales["Order_Status"], "Order_Date": order_dates})
status_dates.groupby("Status")["Order_Date"].agg(["min", "max"])

,min,max
Status,,
Cancelled,2024-06-01,2026-06-30
Delivered,2024-06-01,2026-06-30
Processing,2024-06-01,2026-06-30
Returned,2024-06-01,2026-06-30
Shipped,2024-06-01,2026-06-30


In [101]:
order_dates = pd.to_datetime(sales["Order_Date"], errors="coerce", dayfirst=True)

## 8. How the Tables Connect (Key Checks)

In [102]:
print("Order_ID unique in sales:", sales["Order_ID"].is_unique)
print("Customer_ID unique in customers:", customers["Customer_ID"].is_unique)
print("Product_ID unique in products:", products["Product_ID"].is_unique)

Order_ID unique in sales: True
Customer_ID unique in customers: True
Product_ID unique in products: True


In [103]:
missing_customers = ~sales["Customer_ID"].isin(customers["Customer_ID"])
missing_products = ~sales["Product_ID"].isin(products["Product_ID"])

print("Sales with unknown customer:", missing_customers.sum())
print("Sales with unknown product:", missing_products.sum())

Sales with unknown customer: 0
Sales with unknown product: 0


In [104]:
sales["Order_Date"].head(3)

0    2026-06-07
1    2025-02-04
2    2026-03-12
Name: Order_Date, dtype: str

**Observation (key checks):**
- Order_ID is unique, so each row represents one order.
- Customer_ID and Product_ID are unique in their lookup tables, so joins will not duplicate rows.
- No orphan records: every sale matches an existing customer and product. The tables can be joined safely.

## 9. Do the Money Columns Add Up?

In [105]:
calc_order_value = sales["Quantity"] * sales["Unit_Price"]
calc_total = sales["Order_Value"] + sales["Shipping_Cost"] - sales["Coupon_Discount"]

print("Order_Value mismatches:", (~np.isclose(calc_order_value, sales["Order_Value"])).sum())
print("Total_Amount mismatches:", (~np.isclose(calc_total, sales["Total_Amount"])).sum())

Order_Value mismatches: 0
Total_Amount mismatches: 0


## 10. Testing My Hypotheses About Missing Values

In [106]:
no_coupon = sales["Coupon_Code"].isnull()
print("Orders without coupon:", no_coupon.sum())
print("Of those, discount is 0:", (sales.loc[no_coupon, "Coupon_Discount"] == 0).sum())

Orders without coupon: 199815
Of those, discount is 0: 199815


In [107]:
pd.crosstab(sales["Order_Status"], sales["Rating"].notnull())

Rating,False,True
Order_Status,,
Cancelled,12507,0
Delivered,80109,120030
Processing,12402,0
Returned,12493,0
Shipped,12459,0


## Summary of Data Understanding

**Data overview**
- sales: 250,000 orders (one row = one order), customers: 40,000, products: 2,000.
- Period: 1 June 2024 – 30 June 2026 (25 months, including two Diwali seasons).
- Coverage: 10 states and 39 cities only.

**Data quality**
- No duplicate rows. All IDs are unique, and there are no orphan records, so the tables can be joined safely.
- Money formulas verified: Order_Value = Quantity × Unit_Price; Total_Amount = Order_Value + Shipping_Cost − Coupon_Discount.
- Coupon_Code blanks (80%) = no coupon used (verified: discount is 0 in all of them).
- Rating/Review blanks (52%) are structural: only delivered orders are rated; ~60% of delivered orders have a rating.
- 5 orders have a negative Total_Amount, caused by flat coupons on items under ₹40.

**Business rules discovered**
- Free shipping on orders above ~₹500 (no overlap between paid and free shipping around this value).
- Flat coupons (DIWALI100, FLAT50) have no minimum order value, creating loss-making orders.

**Metric definitions**
- GMV = Total_Amount of all orders. Net Revenue = Total_Amount of Delivered orders only.
- Processing/Shipped orders are treated as "Incomplete".

**Cleaning to-do list (Notebook 02)**
1. Convert Order_Date, Delivery_Date (sales) and Registration_Date, Date_of_Birth (customers) to dates.
2. Fill Coupon_Code blanks with "No Coupon".
3. Leave Rating and Review_Text blanks as they are.
4. Cap negative Total_Amount at 0 and add a flag column.
5. Rename "UP" to "Uttar Pradesh".
6. Create new columns: Year_Month, Weekday, Order_Hour, Delivery_Days (Delivered only), Has_Coupon, Free_Shipping.
7. Join product and customer details onto sales.
8. Save the cleaned file to data/processed/.

**Limitations (simulated data)**
- Non-delivered statuses are each ~5% and spread evenly over time.
- Ratings range only 3–5 (no low ratings).
- The 65+ age group is small (~2% of orders).